mautic/mautic
# Email-Category unsubscribe: Fixing an issue where contacts who have unsubscribed from a category will still receive emails in that category if they match the segment.

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [Yes]
| New feature/enhancement? (use the a.x branch)      | [4.x]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [Yes] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 



#### Description:

An issue where contacts who have unsubscribed from a category will still receive emails in that category if they match the segment.

#### Steps to reproduce the bug:

1. Create 2 (or more) global categories
2. On your test contact record, ensure that the contact has both categories selected/available
3. Create a preference center page that includes the categories block
4. Create an email with that preference center attached
5. Create a segment with your contact
6. Send the email to that segment
7. From the email (in a different browser, to be safe), click the unsubscribe link to get to the preference center
8. Deselect one of the categories and save preferences. Be sure that Email is still selected as a channel you can receive messages on
9. Check the contact record to see if the deselected category is no longer available
10. Create another email, and assign the deselected category to that email
11. Send the email to the same contact
12. Check the contact’s preferences again.

#### Steps to test this PR:

1. Load up [this PR](https://mautibox.com)
2. Create 2 (or more) global categories
3. Create Contacts and update their category preferences
    1. To do so, navigate to the contact page
    2. Navigate to the contact detail page and select preference from the context menu
    3. In the category tab, select all that apply and Save.
    4. Redo the above steps to unsubscribe from the category.
4. Create a segment with your contact; the segment should have the mix of leads subscribed/ unsubscribed category
5. Create an email
    1. Segment email
        1. Create a segment type email with
            1. Selecting created segment
            2. Select the category that the user unsubscribed from
        2. Schedule the email
        3. Check the
    2. For Campaign Action: Sent Email
        1. Create a template email
            1. Selecting created segment
            2. Selecting Preference center
        2. Create a campaign with Segment
        3. Add Action *Send email* selecting the above template email
        4. Save and trigger the campaign.

#### List of areas covered by the unit and/or functional tests:

1. Functional test cases covering new changes:
    1. `\Mautic\EmailBundle\Entity\EmailRepository::getEmailPendingQuery`
    2. `\Mautic\LeadBundle\Entity\LeadCategoryRepository::getLeadCategories`
    3. `\Mautic\LeadBundle\Model\LeadModel::setFrequencyRules`
    4. `\Mautic\EmailBundle\EventListener\CampaignSubscriber::onCampaignTriggerActionSendEmailToContact`
2. Create `\Mautic\CoreBundle\Tests\Functional\CreateTestEntitiesTrait` trait class for creating different entities to avoid duplication in the function testcases.



<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10879"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
